### PR TITLE
amazon-image: amazon-init script checks for nix expressions in userdata correctly

### PR DESCRIPTION
**Pull request by PsyanticY**
A comment in [amzon-init](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/virtualisation/amazon-init.nix) says that it will check for valid nix expressions but it do not do that. when lunching a nixos image from aws and typing whatever in the user-data field the amazon-init unit will copy that to configuration.nix and then it will fail when it run `nixos-rebuild switch` .

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by PsyanticY** [comment]
@matthewbauer  Any thoughts on this ^^

File: nixos/modules/virtualisation/amazon-init.nix [hunk 1]
```diff
@@ -26,12 +26,14 @@ let
...
-        cp "$userData" /etc/nixos/configuration.nix
-      else
-        echo "user data does not appear to be a Nix expression; ignoring"
-        exit
+        nix-instantiate --eval-only "$userData" &> /dev/null
+        if [ $? -eq 0 ]; then
+          echo "setting configuration from EC2 user data"
+          cp "$userData" /etc/nixos/configuration.nix
+        else
+          echo "user data does not appear to be a Nix expression; ignoring"
```
Comment (veprbl) [inline nixos/modules/virtualisation/amazon-init.nix]: You should probably change wording of this to indicate that the expression failed to evaluate.

Comment (PsyanticY) [reply nixos/modules/virtualisation/amazon-init.nix]: @veprbl  thanks for the review. Would `Failed to evaluate user data with nix-instantiate; ignoring` be a more suitable message.

Comment (veprbl) [reply nixos/modules/virtualisation/amazon-init.nix]: Ok, maybe, evaluate wasn't a right word here. How about "user data is not a valid Nix expression; ignoring"?

Comment (PsyanticY) [reply nixos/modules/virtualisation/amazon-init.nix]: @veprbl yeah thats wayy better.

**Comment by veprbl** [comment]
One other comment I have is that after your change, the else branch of
```
if sed '/^\(#\|SSH_HOST_.*\)/d' < "$userData" | grep -q '\S'; then
```
doesn't have a warning and, more importantly, it doesn't terminate with `exit`.

**Comment by edolstra** [comment]
I don't see why this is necessary. The Nix expression will be checked eventually anyway by nixos-rebuild...

**Comment by PsyanticY** [comment]
@veprbl Added the else clause as advised.
@edolstra It is definitely not necessary but it can prevent copying wrong expressions to `/etc/nixos/configuration.nix`
Proving `this is whatever`  as user data results in this
```
[root@ip-10-100-46-187:~]# cat /etc/nixos/configuration.nix 
this is whatever
[root@ip-10-100-46-187:~]# 
```

**Comment by edolstra** [comment]
Well, I don't see why it matters whether it fails in amazon-init or in `nixos-rebuild`. Either way your configuration doesn't work. Also, `nix-instantiate --eval` is a very shallow test (especially on NixOS modules) - it doesn't actually guarantee that the system configuration works.

**Comment by PsyanticY** [comment]
@edolstra mmmm, yeah. Closing then.

**Pull request closed without merging**